zkfacade/src/main/java/com/yahoo/vespa/curator/SingletonManager.java

```diff
@@ -0,0 +1,331 @@
+package com.yahoo.vespa.curator;
+
+import com.yahoo.path.Path;
+import com.yahoo.protect.Process;
+import com.yahoo.vespa.curator.api.VespaCurator.SingletonWorker;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Manages {@link com.yahoo.vespa.curator.api.VespaCurator.SingletonWorker}.
+ *
+ * @author jonmv
+ */
+class SingletonManager implements AutoCloseable {
+
+    private static final Logger logger = Logger.getLogger(SingletonManager.class.getName());
+
+    private final Curator curator;
+    private final Clock clock;
+    private final Duration tickTimeout;
+    private final Map<String, Janitor> janitors = new HashMap<>();
+    private final Map<String, Integer> count = new HashMap<>();
+    private final Map<SingletonWorker, String> registrations = new HashMap<>();
+
+    SingletonManager(Curator curator, Clock clock, Duration tickTimeout) {
+        this.curator = curator;
+        this.clock = clock;
+        this.tickTimeout = tickTimeout;
+    }
+
+    synchronized CompletableFuture<?> register(String singletonId, SingletonWorker singleton) {
+        if (singletonId.contains("/") || singletonId.contains("..")) {
+            throw new IllegalArgumentException("singleton ID may not contain '/' or '..', but got " + singletonId);
```

Comment:
Why this restriction?

Reply:
Because the ID translates to a ZK path, used for locking.

Reply:
The ID should also be non-empty then.

Reply:
Good point.